Set up rust

steveklabnik opened the pull request:
Here's the next chapter! We finally have Rust code!

steveklabnik committed:
working towards a rust kmain

steveklabnik committed:
kmain: it works

steveklabnik committed:
Split this chapter into bits

And move the install instructions out, and fix them up.

eddyb commented on src/hello-from-rust.md:
````diff
@@ -0,0 +1,186 @@
+```
+
+This is our main function, which is traditionally called `kmain()`, for 'kernel
+main.' We need to use the `#[no_mangle] and `pub extern` annotations to indicate
````
Missing backtick after `#[no_mangle]`.

steveklabnik replied:
thanks!

bjorn3 commented on src/creating-our-first-crate.md:
````diff
@@ -0,0 +1,461 @@
+
+```bash
+$ make
+$ make
````
Duplicate line.

steveklabnik replied:
thanks!

bjorn3 commented on src/hello-from-rust.md:
```diff
@@ -0,0 +1,186 @@
+
+This is our main function, which is traditionally called `kmain()`, for 'kernel
+main.' We need to use the `#[no_mangle] and `pub extern` annotations to indicate
+that we're going to call this function like we would call a C function. The `->`
```
-> !

steveklabnik replied:
on it

steveklabnik committed:
nits

bjorn3 commented on src/hello-from-rust.md:
```diff
@@ -0,0 +1,186 @@
+This is our main function, which is traditionally called `kmain()`, for 'kernel
+main.' We need to use the `#[no_mangle]` and `pub extern` annotations to indicate
+that we're going to call this function like we would call a C function. The `-> !`
+indicates that this function never returns. And in fact, it does not: the body
```
Incorrect not in `it does not`.

steveklabnik replied:
Hmmm, I think this is one of things things in which English is ambiguous. 

> And in fact, it does not return

is how I read this, but you're reading it as negating the "never returns."

bjorn3 replied:
Yes, could you change it to `it does not return`?

justinvoss commented on src/creating-our-first-crate.md:
````diff
@@ -0,0 +1,460 @@
+$ cd ../..
+```
+
+Whew! That'sa  lot of work. Don't worry, we'll be modifying our `Makefile` to do this
````
Spacing between "That's" and "a" is off.

guewen commented on src/creating-our-first-crate.md:
```diff
@@ -0,0 +1,460 @@
+
+The rest of these options aren't particularly interesting. I would tell you to go
+look them up in Rust's documentation, but it's sorely lacking at the moment. Maybe
+I should stop writing this and go work on that... anway. I digress.
```
s/anway/anyway/

guewen commented on src/creating-our-first-crate.md:
```diff
@@ -0,0 +1,460 @@
+defines most of them, but the last two, `panic_fmt` and `eh_personality`,
+need to be defined by us.
+
+Both of these language items invovle a feature of Rust called 'panics.'
```
s/invovle/involve/

guewen commented on src/hello-from-rust.md:
```diff
@@ -0,0 +1,186 @@
+In this case, the reason we need `unsafe` is the next two lines: we create a
+pointer to `0xb8000`, and then write some numbers to it. Rust cannot know that
+this is safe; if it did, it would have to understand that we are a kernel,
+and understand the VGA specification. Having a programming langauge understand
```
s/langauge/language/

steveklabnik replied:
i am the worst. thanks for finding these :smile:

steveklabnik committed:
fix spelling nits

steveklabnik committed:
Show off alternate Makefile

steveklabnik commented:
Thanks everyone, I've addressed the nits. Also, I wrote an aside pointing to an example of a more DRY makefile.

steveklabnik committed:
Be actually as not-strict as we say

as discovered in https://github.com/intermezzOS/kernel/pull/45, the
focus on the .a wouldn't actually cause things to be rebuilt. This will.

steveklabnik commented:
I am too excited to get this merged, and so I'm going to go ahead and do it. If you didn't get the chance to check this chapter out yet, please open issues with anything you want to comment on! Still very happy to get feedback.
